File: frameworks/tf/layer_norm.py
# Copyright 2020 LMNT, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

"""Layer Normalization"""


import pkg_resources
import tensorflow as tf

from tensorflow.compat import v1


__all__ = [
    'LayerNorm'
]


LIB = tf.load_op_library(pkg_resources.resource_filename(__name__, 'libhaste_tf.so'))


@tf.RegisterGradient("HasteLayerNorm")
def layer_norm_gradient(op, *grads):
  x = op.inputs[0]
  gamma = op.inputs[1]
  beta = op.inputs[2]
  cache = op.outputs[1]

  return LIB.haste_layer_norm_grad(x, gamma, beta, grads[0], cache)


class LayerNorm(tf.Module):
  """
  Layer normalization layer.

  This class exposes a fused and GPU-accelerated implementation of layer
  normalization as described by [Ba et al.](https://arxiv.org/abs/1607.06450)
  """

  def __init__(self, name=None):
    """
    Initialize the parameters of the layer normalization layer.

    Arguments:
      name: (optional) string, the name for this layer.
    """
    super(LayerNorm, self).__init__(name)
    self.realname = name
    self.gamma = None
    self.beta = None
    self.built = False

  def build(self, shape):
    """
    Creates the variables of the layer.

    Calling this method is optional for users of the LayerNorm class. It is
    called internally with the correct shape when `__call__` is invoked.

    Arguments:
      shape: instance of `TensorShape`.
    """
    if self.built:
      return
    hidden_size = int(shape[-1])
    with self.name_scope, v1.variable_scope(self.realname, 'layer_norm'):
      self.gamma = v1.get_variable('gamma', shape=[hidden_size], initializer=v1.initializers.ones())
      self.beta = v1.get_variable('beta', shape=[hidden_size], initializer=v1.initializers.zeros())
    self.built = True

  def __call__(self, x):
    """
    Runs the layer.

    Arguments:
      x: Tensor, a rank R tensor.

    Returns:
      y: Tensor, a rank R tensor with the last dimension normalized.
    """
    self.build(x.shape)
    y, _ = LIB.haste_layer_norm(x, self.gamma, self.beta)
    return y

File: frameworks/tf/layer_norm.cc
// Copyright 2020 LMNT, Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

#include "haste.h"
#include "support.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/util/stream_executor_util.h"
#include "tensorflow/stream_executor/stream.h"

using namespace tensorflow;

using haste::v0::layer_norm::ForwardPass;
using haste::v0::layer_norm::BackwardPass;
using tensorflow::se::Stream;
using tensorflow::shape_inference::DimensionHandle;
using tensorflow::shape_inference::InferenceContext;
using tensorflow::shape_inference::ShapeHandle;

REGISTER_OP("HasteLayerNorm")
    .Attr("R: {float, double}")
    .Input("x: R")
    .Input("gamma: R")
    .Input("beta: R")
    .Output("y: R")
    .Output("cache: R")
    .SetShapeFn([](InferenceContext* c) {
      ShapeHandle input_shape;
      ShapeHandle gamma_shape;
      ShapeHandle beta_shape;

      TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 2, &input_shape));
      TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 1, &gamma_shape));
      TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 1, &beta_shape));

      c->set_output(0, input_shape);
      c->set_output(1, c->UnknownShapeOfRank(2));
      return Status::OK();
    });

template<typename T>
struct HasteLayerNormOp : public OpKernel {
  explicit HasteLayerNormOp(OpKernelConstruction* context) : OpKernel(context) {}

  void Compute(OpKernelContext* context) override {
    const Tensor& x = context->input(0);
    const Tensor& gamma = context->input(1);
    const Tensor& beta = context->input(2);

    const auto batch_size = x.shape().dim_size(0);
    const auto hidden_size = x.shape().dim_size(1);

    Tensor* y = nullptr;
    OP_REQUIRES_OK(context, context->allocate_output(0, x.shape(), &y));

    Tensor* cache = nullptr;
    OP_REQUIRES_OK(context, context->allocate_output(1, { batch_size, 2 }, &cache));

    ForwardPass<T> forward(
        batch_size,
        hidden_size,
        gamma.flat<T>().data(),
        beta.shape().dim_size(0) ? beta.flat<T>().data() : nullptr,
        cache->flat<T>().data());

    forward.Run(GetCudaStream(context), x.flat<T>().data(), y->flat<T>().data());
  }
};

REGISTER_GPU_KERNEL(HasteLayerNorm, float);
REGISTER_GPU_KERNEL(HasteLayerNorm, double);

REGISTER_OP("HasteLayerNormGrad")
    .Attr("R: {float, double}")
    .Input("x: R")
    .Input("gamma: R")
    .Input("beta: R")
    .Input("dy: R")
    .Input("cache: R")
    .Output("dx: R")
    .Output("dgamma: R")
    .Output("dbeta: R")
    .SetShapeFn([](InferenceContext* c) {
      ShapeHandle x_shape;
      ShapeHandle gamma_shape;
      ShapeHandle beta_shape;
      ShapeHandle dy_shape;
      ShapeHandle cache_shape;

      TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 2, &x_shape));
      TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 1, &gamma_shape));
      TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 1, &beta_shape));
      TF_RETURN_IF_ERROR(c->WithRank(c->input(3), 2, &dy_shape));
      TF_RETURN_IF_ERROR(c->WithRank(c->input(4), 2, &cache_shape));

      c->set_output(0, x_shape);
      c->set_output(1, gamma_shape);
      c->set_output(2, beta_shape);
      return Status::OK();
    });

template<typename T>
struct HasteLayerNormGradOp : public OpKernel {
  explicit HasteLayerNormGradOp(OpKernelConstruction* context) : OpKernel(context) {}

  void Compute(OpKernelContext* context) override {
    const Tensor& x = context->input(0);
    const Tensor& gamma = context->input(1);
    const Tensor& beta = context->input(2);
    const Tensor& dy = context->input(3);
    const Tensor& cache = context->input(4);

    const auto batch_size = x.shape().dim_size(0);
    const auto hidden_size = x.shape().dim_size(1);
    const auto cache_shape = context->input(4).shape();
    const auto data_type = DataTypeToEnum<T>::value;

    Tensor* dx = nullptr;
    OP_REQUIRES_OK(context, context->allocate_output(0, x.shape(), &dx));

    Tensor* dgamma = nullptr;
    OP_REQUIRES_OK(context, context->allocate_output(1, gamma.shape(), &dgamma));

    Tensor* dbeta = nullptr;
    OP_REQUIRES_OK(context, context->allocate_output(2, beta.shape(), &dbeta));

    cudaMemset(dgamma->flat<T>().data(), 0, dgamma->AllocatedBytes());
    cudaMemset(dbeta->flat<T>().data(), 0, dbeta->AllocatedBytes());

    BackwardPass<T> backward(
        batch_size,
        hidden_size,
        gamma.flat<T>().data(),
        beta.shape().dim_size(0) ? beta.flat<T>().data() : nullptr,
        x.flat<T>().data(),
        dgamma->flat<T>().data(),
        beta.shape().dim_size(0) ? dbeta->flat<T>().data() : nullptr,
        const_cast<T*>(cache.flat<T>().data()));

    backward.Run(GetCudaStream(context), dy.flat<T>().data(), dx->flat<T>().data());
  }
};

REGISTER_GPU_KERNEL(HasteLayerNormGrad, float);
REGISTER_GPU_KERNEL(HasteLayerNormGrad, double);
